messaging/src/main/java/org/axonframework/messaging/annotation/MessageHandlerInterceptorDefinition.java

```diff
@@ -0,0 +1,123 @@
+
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * MethodEnhancerDefinition that marks methods (meta-)annotated with {@link MessageHandlerInterceptor}
+ * as interceptors. These methods need to be given special treatment when invoking handlers. Matching
+ * interceptors need to be invoked first, allowing them to proceed the invocation chain.
+ * <p>
+ * This definition also recognizes interceptors only acting on the response. These must be meta-annotated with
+ * {@link ResultHandler}.
+ */
+public class MessageHandlerInterceptorDefinition implements HandlerEnhancerDefinition {
+    @Override
+    public <T> MessageHandlingMember<T> wrapHandler(MessageHandlingMember<T> original) {
+
+        if (original.annotationAttributes(MessageHandlerInterceptor.class).isPresent()) {
+            Optional<Map<String, Object>> attributes = original.annotationAttributes(ResultHandler.class);
+            if (attributes.isPresent()) {
+                return new ResultHandlingInterceptorMember<>(original, (Class<?>) attributes.get().get("resultType"));
+            }
+            return new InterceptedMessageHandlingMember<>(original);
+        }
+        return original;
+    }
+
+    private static class ResultHandlingInterceptorMember<T> extends WrappedMessageHandlingMember<T> implements MessageHandlerInterceptorMember<T> {
+
+        private final Class<?> expectedResultType;
+
+        public ResultHandlingInterceptorMember(MessageHandlingMember<T> original, Class<?> expectedResultType) {
+            super(original);
+            this.expectedResultType = expectedResultType;
+            Method method = original.unwrap(Method.class).orElseThrow(() -> new AxonConfigurationException(
+                    "The @ExceptionHandler must be on method."));
+            boolean declaredInterceptorChain = Arrays.stream(method.getParameters())
+                                                     .anyMatch(p -> p.getType().equals(InterceptorChain.class));
+            if (declaredInterceptorChain) {
+                throw new AxonConfigurationException("@ExceptionHandler annotated methods should not delare an InterceptorChain parameter");
+            }
+        }
+
+        @Override
+        public boolean canHandle(Message<?> message) {
+            return ResultParameterResolverFactory.ignoringResultParameters(() -> super.canHandle(message));
+        }
+
+        @Override
+        public Object handle(Message<?> message, T target) throws Exception {
+            InterceptorChain chain = InterceptorChainParameterResolverFactory.currentInterceptorChain();
+            try {
+                return chain.proceed();
+            } catch (Exception e) {
+                if (!expectedResultType.isInstance(e)) {
+                    throw e;
+                }
+                return ResultParameterResolverFactory.callWithResult(e, () -> {
+                    if (super.canHandle(message)) {
```

Comment:
Isn't this `super.canHandle` check already performed in the local `canHandle` call? What's the scenario when this is false so that the exception needs to be thrown again?

Reply:
No, `canHandle` has to be permissive (it ignores any parameters that expect an invocation result), since we only know if the can _really_ handle this message when we also know the result. In this case, we also explicitly check if the result is of the type the handler expects.

Reply:
Aaah gotcha, of course!